### DISCUSSION TITLE
Using a multi-stage build to add micromamba functionality to an existing image

### DISCUSSION THREAD

**Discussion by wholtz** [mamba-org/micromamba-docker#175] [Show and tell]
[Dockerfile for tomcat with micromamba](https://github.com/biorack/labkey_deploy/blob/9636bea664f4291c0cd1a3e42877485d4b45a691/labkey/Dockerfile)

I needed to use a tomcat image but wanted to have multiple R/Python (conda) environments within the final image.  I could have copied and pasted the contents of the micromamba `Dockerfile` into a new `Dockerfile` that started with `FROM tomcat`, but then I'd need to copy all the scripts the micromamba `Dockerfile` calls into my new repo as well. If any of those scripts were later modified within the `mamba-org/micromamba-docker` repo, I would then have to manually update my downstream repo to keep up to date with the changes. By using a multi-stage build, I am able to copy the scripts directly from the `mambaorg/micromamba` image.  Now if an updated version of `mambaorg/micromamba` is released, I only need to update the `mambaorg/micromamba` tag within my downstream `Dockerfile` and it will pickup the newer `micromamba` binary along with any changes to the included scripts.

The exercise has me debating with myself if the following code from the micromamba `Dockerfile` should be moved into a separate script file:
https://github.com/mamba-org/micromamba-docker/blob/41d5bd8c2b70a5d441b657a096199a9271f256ed/Dockerfile#L34-L41

The argument for such a change, is that my above use case would then automatically track changes to this code block when I update the `mambaorg/micromamba` tag within my downstream `Dockerfile`. A reason for not making this change is that it hides details in other files which makes it hard for someone to glance at the `Dockerfile` and understand exactly what is being done.

The functionality in the above code block can be broken into 2 topics:
1. user account creation
1. setup of the directory `$MAMBA_ROOT_PREFIX`

Perhaps it makes sense to create a script for each of these two activities. This should give a reader of the  `mamba-org/micromamba-docker`  `Dockerfile` a better  conceptual understanding of what is being done, even if the details require going into other files.

@maresb - care to offer an opinion on this possible change?

**Comment by maresb** [comment]
> care to offer an opinion on this possible change?

I'm all in favor of multiple small scripts with long descriptive filenames, e.g. along the lines of `_dockerfile_initialize_user_accounts.sh` and `_dockerfile_setup_root_prefix.sh`. This seems to me like the best of both worlds: the Dockerfile remains concise and readable, and the interested reader doesn't need to go far to find the details if desired.

(In general I find the problem of "merging" Dockerfiles intriguing. It seems to be a common problem with no good solution. I would have suggested modifying tomcat to use micromamba as the base, until I noticed that tomcat uses eclipse as a base image, which is an annoying complication.)

**Reply by wholtz** [reply]
The docker developers seem to consider multi-stage builds to be their answer to merging Dockerfiles - https://github.com/moby/moby/issues/24445. But it does seem better and more powerful solutions were possible.

**Reply by maresb** [reply]
Thanks for the reference! That sounds like an interesting proposal, but I imagine it could easily read to a nightmare of subtle bugs in terms of data consistency with what would effectively become asynchronous filesystem writes.

**Reply by wholtz** [reply]
The proposed additional shell script files have been implemented in 9a608bf9252770c5ca72d8d42c4e64553899e385. Supporting documentation has been added as well: https://github.com/mamba-org/micromamba-docker#adding-micromamba-to-an-existing-docker-image

**Reply by maresb** [reply]
Very nicely done, as usual! :smile:
